Migrate airflow-database guide (#1081)
astronomer/docs: guides/airflow-database.md

---
title: "Understanding the Airflow metadata database"

> sidebar_label: "Understanding the Airflow metadata database"

Reviewer: ```suggestion
sidebar_label: "Metadata database"
```

> description: "A walkthrough of the Apache Airflow metadata database, with a full ERD."

Reviewer: ```suggestion
description: "Learn about everything you need to use the Apache Airflow metadata database."
```

id: airflow-database

> tags: ["Database", "SQL", "Components"]

Reviewer: ```suggestion
tags: [Database, SQL, Components]
```

---

The metadata database is a core component of Airflow. It stores crucial information such as the configuration of your Airflow environment's roles and permissions, as well as all metadata for past and present DAG and task runs.

A healthy metadata database is critical for your Airflow environment. Losing data stored in the metadata database can both interfere with running DAGs and prevent you from accessing data for past DAG runs. As with any core Airflow component, having a backup and disaster recovery plan in place for the metadata database is essential.

In this guide, you'll learn everything you need to know about the Airflow metadata database to ensure a healthy Airflow environment, including:

- Database specifications.
- Important content stored in the database.
- Best practices for using the metadata database.
- Different methods for accessing data of interest.

> > **Note**: Astronomer strongly advise against directly modifying the metadata database, as this can cause dependency issues and corrupt your Airflow instance.

Reviewer: ```suggestion
```

Changes to the Airflow metadata database configuration and its schema are very common and happen with almost every minor update. For this reason, prior to Airflow 2.3 you should not downgrade your Airflow instance in place. With Airflow 2.3 the `db downgrade` command was added, providing an option to [downgrade Airflow](https://airflow.apache.org/docs/apache-airflow/2.3.0/usage-cli.html#downgrading-airflow).

> **Note**: Always backup your database before running any database operations!

## Metadata database content

> **Note**: For many use cases you can access contents from the metadata database in the Airflow UI or the [stable REST API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html). These points of access always beat querying the metadata database directly!

There are several types of metadata stored in the metadata database.

- User login information and permissions.
- Information used in DAGs, like variables, connections and XComs.
- Data about DAG and task runs which are generated by the scheduler.
- Other minor tables, such as tables which store DAG code in different formats or information about import errors.

> Most of this data can be queried using the [Airflow REST API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html).

Reviewer: ```suggestion
For many use cases you can access content from the metadata database in the Airflow UI or the [stable REST API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html). These points of access always beat querying the metadata database directly!
```

- **Audit logs** shows events of various types that were logged to the metadata database (for example, DAGs being paused or tasks being run).
- **Task Instances** contains a record of every task run with a variety of attributes such as the priority weight, duration, or the URL to the task log.
- **Task Reschedule** lists tasks that have been rescheduled.
- **Triggers** shows all currently running [triggers](https://www.astronomer.io/guides/deferrable-operators).
- **SLA Misses** keeps track of tasks that missed their [SLA](https://www.astronomer.io/guides/error-notifications-in-airflow/#airflow-slas).

### Other tables

There are additional tables in the metadata database storing data ranging from DAG tags over serialized DAG code, import errors to current states of sensors. Some of the information in these tables will be visible in the Airflow UI in various places:

- The source code of DAGs can be found by clicking on a DAG name from the main view and then going to the **Code** view.
- Import errors appear at the top of the **DAGs** view in the UI.
- DAG tags will appear underneath their respective DAG with a cyan background.

## Airflow metadata database best practices

> 1. When upgrading or downgrading Airflow, always follow the [recommended steps for changing Airflow versions](https://airflow.apache.org/docs/apache-airflow/stable/installation/upgrading.html?highlight=upgrade): back up the metadata database, check for deprecated features, pause all DAGs, and make sure no tasks are running.

Author: @jwitz - done!

```python
# import the request library
import requests
import os

# provide the location of your airflow instance
ENDPOINT_URL = "http://localhost:8080/"

# in this example env variables were used to store login information
# you will need to provide your own credentials
user_name = os.environ['USERNAME_AIRFLOW_INSTANCE']
password = os.environ['PASSWORD_AIRFLOW_INSTANCE']

# query the API for successful task instances from all dags and all dag runs (~)
req = requests.get(
  f"{ENDPOINT_URL}/api/v1/dags/~/dagRuns/~/taskInstances?state=success",  
  auth=(user_name, password))

# from the API response print the value for "total entries"
print(req.json()['total_entries'])
```

It is also possible to navigate to **Browse** -> **Task Instances** in the Airflow UI and filter the task instances for all with a state of `success`. The `Record Count` will be on the right side of your screen.

![Count successful tasks Airflow UI](/img/guides/successful_tasks_UI.png)

> ### Example: Pause and unpause a DAG

Reviewer: ```suggestion
### Pause and unpause a DAG
```

The Python script below sends a PATCH request to the Airflow API to update the entry for the DAG with a specific ID (here `example_dag_basic`), which is paused (`update_mask=is_paused`) with a json that will set the `is_paused` property to `True` therefore unpausing the DAG.

```python
# import the request library
import requests
import os

# provide the location of your airflow instance
ENDPOINT_URL = "http://localhost:8080/"

# in this example env variables were used to store login information
# you will need to provide your own credentials
user_name = os.environ['USERNAME_AIRFLOW_INSTANCE']
password = os.environ['PASSWORD_AIRFLOW_INSTANCE']

# data to update, for unpausing, simply set this to False
update= {"is_paused": True}
# specify the dag to pause/unpause
dag_id = example_dag_basic

# query the API to patch all tasks as paused
req = requests.patch(
  f"{ENDPOINT_URL}/api/v1/dags/{dag_id}?update_mask=is_paused", json=update,
  auth=(user_name, password))

# print the API response
print(req.text)
```

> ### Example: Delete a DAG

Reviewer: ```suggestion
### Delete a DAG
```

Deleting the metadata of a DAG can be accomplished either by clicking the `trashcan` icon in the Airflow UI or sending a `DELETE` request with the Airflow REST API. This is not possible while the DAG is still running, and will not delete the Python file in which the DAG is defined, meaning the DAG will appear again in your UI with no history at the next parsing of the `/dags` folder from the scheduler.

The Python script below sends a DELETE request to a DAG with a specific ID (here: `dag_to_delete`).

```python
# import the request library
import requests
import os

# provide the location of your airflow instance
ENDPOINT_URL = "http://localhost:8080/"

# in this example env variables were used to store login information
# you will need to provide your own credentials
user_name = os.environ['USERNAME_AIRFLOW_INSTANCE']
password = os.environ['PASSWORD_AIRFLOW_INSTANCE']

# specify which dag to delete
dag_id = 'dag_to_delete'

# send the deletion request
req = requests.delete(
  f"{ENDPOINT_URL}/api/v1/dags/{dag_id}",
  auth=(user_name, password))

# print the API response
print(req.text)
```

> ### Example: Retrieve all DAG dependencies

Reviewer: ```suggestion
### Retrieve all DAG dependencies
```

To programmatically access this information, you can use SQLAlchemy with [Airflow models](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/index.html) to access data from the metadata database. Note that if you are running Airflow in a Dockerized setting, you have to run the script below from within your scheduler container.

```python
from sqlalchemy import create_engine
from sqlalchemy.orm import Session
from airflow.models.serialized_dag import SerializedDagModel
import os

# retrieving your SQL Alchemy connection
# if you are using Astro CLI this env variable will be set up automatically
sql_alchemy_conn = os.environ['AIRFLOW__CORE__SQL_ALCHEMY_CONN']

conn_url = f'{sql_alchemy_conn}/postgres'

engine = create_engine(conn_url)

with Session(engine) as session:
    result = session.query(SerializedDagModel).first()
    print(result.get_dag_dependencies())
```

> ### Example: Retrieve Alembic version

Reviewer: ```suggestion
### Retrieve alembic version
```
